hermes-client/src/main/java/pl/allegro/tech/hermes/client/HermesClient.java

```diff
@@ -67,16 +71,41 @@ private String createUri(URI uri) {
     }
 
     public CompletableFuture<HermesResponse> publish(HermesMessage message) {
+        if (shutdown) {
+            return completedWithShutdownException();
+        }
         HermesMessage.appendDefaults(message, defaultHeaders);
         return publish(message, (response) -> retryCondition.test(response) ? sendOnce(message) : completedFuture(response));
     }
 
     private CompletableFuture<HermesResponse> publish(HermesMessage message, Function<HermesResponse, CompletionStage<HermesResponse>> retryDecision) {
-        return range(0, retries).boxed().reduce(sendOnce(message), (future, attempt) -> future.thenCompose(retryDecision), (future, attempt) -> future);
+        currentlySending.incrementAndGet();
+        return range(0, retries).boxed().reduce(sendOnce(message), (future, attempt) -> future.thenCompose(retryDecision), (future, attempt) -> future)
+                .whenComplete((response, ex) -> currentlySending.decrementAndGet());
     }
 
     private CompletableFuture<HermesResponse> sendOnce(HermesMessage message) {
         return sender.send(URI.create(uri + message.getTopic()), message).exceptionally(HermesResponseBuilder::hermesFailureResponse);
     }
 
+    private CompletableFuture<HermesResponse> completedWithShutdownException() {
+        CompletableFuture<HermesResponse> alreadyShutdown = new CompletableFuture<>();
+        alreadyShutdown.completeExceptionally(new HermesClientShutdownException());
+        return alreadyShutdown;
+    }
+
+    public void shutdown() {
```

Comment:
maybe make it closeable then:

```
close() # this will block until it closes with some default timeout
close(int timeout) # this will close in some timeout
CompletableFuture<> closeAsync(int timeout) # this will return future you can block on to close
```